Use non-transient systems instead of transient systems

loganharbour opened the pull request:
Refs #16339

loganharbour committed:
Use non-transient libMesh systems instead of transient systems

Initialize only the solution states that are requested

refs #16339

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/638348/) on e37f2de wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/16363/site)

This comment will be updated on new commits.

loganharbour commented:
This is a pretty significant change - so I'd prefer a few eyes if possible.

loganharbour commented:
By default, now, only the current solution state is initialized.

Older states are initialized by:
- Calling the non-const `System::solutionOld()`, `System::solutionOlder()`, or `System::solutionState()` getters. 
- `MooseVariableData` requesting old/older values via kernels, bcs, etc

With this, ONLY the states requested are initialized. Therefore, if someone wants old in aux for example: NL, displaced NL, and aux NL will not have the old state. 

It may be worthwhile restricting the construction of older states after initialSetup()... would like an opinion on this.

In addition, we could likely also improve the time integrators by only requesting old in the base integrator and then requesting older if it's needed in the higher order integrators.

lindsayad commented:
My thinking in #16339 was to go even further than this. I thought that we would even eliminate the system matrix, see https://github.com/libMesh/libmesh/pull/2638#issuecomment-734052222 and https://github.com/idaholab/moose/issues/16325. @YaqiWang was that your thinking as well?

loganharbour commented:
@lindsayad I'm actually trying that right now - I just figured it would be easier to not jump the gun completely and start with the solution.

lindsayad commented:
Sounds good. If we're going to go piecemeal (as the PR title currently indicates), let's just to make sure we're reffing 16339 instead of closing.

YaqiWang commented:
My goal is a clean code with unnecessary vectors removed for steady-state and system matrix removed for JFNK solve type. I suggest we use `System` that @lindsayad mentioned in #16339.

loganharbour commented:
I agree.

The issue is that the switch from `NonlinearImplicitSystem` -> `System` is significant compared to this. I would much prefer to do one at a time for the purposes of reviewing and for debugging.

YaqiWang commented on framework/src/variables/MooseVariableData.C:
```diff
@@ -192,6 +192,17 @@ MooseVariableData<OutputType>::MooseVariableData(const MooseVariableFE<OutputTyp
 
+template <typename OutputType>
+unsigned int
+MooseVariableData<OutputType>::needSolutionStates() const
```
This function can be non-template.

rwcarlsen replied:
It's not a function template - it's a templated class.

YaqiWang commented on framework/src/systems/SystemBase.C:
```diff
@@ -1352,6 +1357,26 @@ SystemBase::name() const
+  unsigned int state = 0;
+
+  // May need older states from the variables (requested in kernels, bcs, etc)
+  for (const auto & var : getVariables(/* tid = */ 0))
```
I was thinking to let the executioner to drive what states we need. Kernels, etc. should error out if required state does not exist. In this way, we can get rid of integrity check currently in executioner. `FEProblemBase::transient` should be set properly based on executioner.

loganharbour replied:
I would be okay with this for the NL system, but not the aux system. There exist cases where you don't need older values in aux but you're running with a transient. In which case, there is no need to init the old aux vectors.

YaqiWang replied:
This sounds to me a further optimization. Typically aux system does not cause any trouble because it is much smaller. I prefer a simple design here, i.e. if NL system has old, let aux system have it too.

lindsayad replied:
I'm inclined to agree with @YaqiWang here. Let the `Executioner` drive

fdkong replied:
>This sounds to me a further optimization. Typically aux system does not cause any trouble because it is much smaller. I prefer a simple design here, i.e. if NL system has old, let aux system have it too.

I just want to mention in NDA, the aux is pretty big

rwcarlsen reviewed (commented):
Too bad this doesn't reduce the amount of code in moose :P  This is a good change - I like it.  The only thing I don't like is that there could be cases that escape our notice where someone wants an old/older solution but we didn't "figure it out" until after it's too late.  This can occur because we don't use the actual retrieval site for the solution[Old/Older] to determine what is needed (indeed this would be "too late" - which is why you didn't implement it that way).  I'm not suggesting we do this, but it might be worth considering:  What if, for the first iteration, we saved everything.  At the system retrievel level for solution old/older, etc. we record exactly what people want/ask for.  And then after the first iteration, we delete/remove any saved vectors/matrices that we don't need.  This check/removal would only have to occur on the first iteration.  Then it would become effectively impossible to miss any cases.

lindsayad commented:
> Too bad this doesn't reduce the amount of code in moose :P This is a good change - I like it. The only thing I don't like is that there could be cases that escape our notice where someone wants an old/older solution but we didn't "figure it out" until after it's too late. This can occur because we don't use the actual retrieval site for the solution[Old/Older] to determine what is needed (indeed this would be "too late" - which is why you didn't implement it that way). I'm not suggesting we do this, but it might be worth considering: What if, for the first iteration, we saved everything. At the system retrievel level for solution old/older, etc. we record exactly what people want/ask for. And then after the first iteration, we delete/remove any saved vectors/matrices that we don't need. This check/removal would only have to occur on the first iteration. Then it would become effectively impossible to miss any cases.

We avoid this I think if we let the `Executioner` determine what vectors/matrices we should have, and erroring out if a user requests if a kernel requests old data that is simply not available according to their `Executioner` type (as @YaqiWang suggested)

lindsayad commented:
I was thinking about this when we were having the `initialSetup` discussion @loganharbour. I'm sorry for not voicing my thought then

rwcarlsen commented:
@lindsayad - I'm saying that we should be able to make it _impossible_ for a user to try to use something that we don't have available - if we use users uses of it as the definition of what we make available.  What I suggested would be an implementation that rather than giving up sometimes with an error saying "sorry - we didn't catch that you used a certain thing and so we threw it away" - we just eliminate that possibility entirely.

lindsayad commented:
>  I'm saying that we should be able to make it _impossible_ for a user to try to use something that we don't have available - if we use users uses of it as the definition of what we make available. What I suggested would be an implementation that rather than giving up sometimes with an error saying "sorry - we didn't catch that you used a certain thing and so we threw it away" - we just eliminate that possibility entirely.

I was thinking you would `mooseError` if someone calls `slnOld` or `slnOlder` (which happens during kernel construction) when the states aren't available. So a hard error up-front

rwcarlsen commented:
I was thinking that we just detect all cases where a user needs something so we don't ever have to abort because we didn't figure out what they needed.

lindsayad commented:
> use users uses

I like this

lindsayad commented:
> And then after the first iteration, we delete/remove any saved vectors/matrices that we don't need.

This would incur memory spikes during initialization that @YaqiWang is working hard to avoid

rwcarlsen commented:
Yeah.  One "middle ground" which might still be undesirable - we could create _only_ the old vec(s) and skip the older[er[er]] ones initially.  Then add the others if needed, or delete the "old" ones if needed.  It just seems a darn shame and sloppy to try to reconstruct what is needed from all the places that we "guessed" users might pull old[er] solutions from 0 instead of just using the ground-zero place that already knows that information as our oracle.

loganharbour commented:
> I was thinking you would mooseError if someone calls slnOld or slnOlder (which happens during kernel construction) when the states aren't available. So a hard error up-front

This is exactly what I'm switching to right now.

lindsayad commented:
> Yeah. One "middle ground" which might still be undesirable - we could create _only_ the old vec(s) and skip the older[er[er]] ones initially. Then add the others if needed, or delete the "old" ones if not needed.

This is a good idea. Most of the time older is not needed. And then in pseudo-transients without time derivative terms, old is also not needed as @rwcarlsen said

lindsayad commented:
> ground-zero place

What is ground-zero? `solutionOld()`

loganharbour commented:
I would prefer that by default we have none (current only). Transient executioner adds old. TimeIntegrators that can produce older values can add older. Couplable checks if the states are available and produces a reasonable error with the requesting object if they're not available. MooseVariableField will error directly if the states are not available.

lindsayad commented:
> I would prefer that by default we have none (current only). Transient executioner adds old. TimeIntegrators that can produce older values can add older. Couplable checks if the states are available and produces a reasonable error with the requesting object if they're not available. MooseVariableField will error directly if the states are not available.

I think this is good. If we want to later we can add the capability to delete the old vector if doing a pseudo-transient.

loganharbour commented:
What about the aux system? I think of two options:
- Initialize them by default with the NL system as they're requested.
- Allow AuxVariables to initialize old solutions as long as the NL system has the same state.

rwcarlsen commented:
See - this is exactly the kind of individual-case-handling I like to generally try to avoid.  Ground zero would be solutionOld, etc. function calls on the system.  Then we would only have to handle things in one place.  But I understand the tradeoff.  It may be desirable to avoid even the old vector creation/deletion memory spike.  You guys can decide whatever you think is better for our current situation.

loganharbour commented:
The driving force for this PR (and the matrix PR to come) is to not init the old vectors in Griffin.

loganharbour commented:
Struggling with dissertation development so I took a break and came back to this for a bit. I think this is in a pretty good shape now.

By default, we only get a state of 0 (current). You obtain older states by the following:

- The non-const versions of `SystemBase::solutionState()`, `SystemBase::solutionOld()`, `SystemBase::solutionOlder()`
- `SystemBase::needSolutionState()` called before `FEProblemBase::initialSetup()`
- Requesting variable data through any of the getters in `MooseVariableData` and also the `Coupleable` interface, within kernels. After init, we look through the variable data and add the additional states that are requested.

The aux system will always have old and older because we request it in `AuxKernel`. There's room for more optimization here in the future. 

I ran tests in `test` to see how this changes our state allocation. For the ~2400 tests that ran, 
- 33% now run with only the old solution (no init of older solution)
- 40% now run with only the current solution (no init of old or older solutions)

loganharbour commented:
Patch is ready for Griffin.

rwcarlsen reviewed (dismissed):
I'll give Yaqi a last chance to chime in before we merge.

fdkong commented on framework/include/variables/MooseVariableScalar.h:
```diff
@@ -141,6 +141,11 @@ class MooseVariableScalar : public MooseVariableBase
+   * The highest solution state required for this variable (0 = current, 1 = old, etc...).
+   */
+  unsigned int needSolutionState() const;
+
```
Could we rename this to `needHighestSolutionState`?

loganharbour replied:
`oldestSolutionState` / `oldestSolutionStateNeeded` ?

fdkong commented on framework/src/systems/SystemBase.C:
```diff
@@ -1328,6 +1333,9 @@ SystemBase::copyOldSolutions()
 void
 SystemBase::restoreSolutions()
 {
+  if (!hasSolutionState(1))
```
I feel like `solution state` should be `enum` instead of `unsigned int` to increase readability.

`enum  SolutionState { current , old, older, ...,  };`

loganharbour replied:
I thought about this too. The problem is that we also use this for state 3, and it could potentially be used for even older states.

In that case - do we have:

- 0: current
- 1: old
- 2: older
- 3: oldoldold/olderer
- 4: oldoldoldold/oldererer

??

lindsayad replied:
`unsigned int` seems fine to me given that it can go arbitrarily high

fdkong commented on framework/src/timeintegrators/CentralDifference.C:
```diff
@@ -32,6 +32,7 @@ validParams<CentralDifference>()
 CentralDifference::CentralDifference(const InputParameters & parameters)
   : ActuallyExplicitEuler(parameters),
     _du_dotdot_du(_sys.duDotDotDu()),
+    _solution_older(_sys.solutionState(2)),
```
If we use `enum`, `2` will become `older`. That help.

loganharbour replied:
See comment above. `solutionOlder()` still works, but I tried to keep things consistent here.

fdkong commented on framework/src/timeintegrators/ExplicitRK2.C:
```diff
@@ -25,7 +25,8 @@ ExplicitRK2::validParams()
   : TimeIntegrator(parameters),
     _stage(1),
-    _residual_old(_nl.addVector("residual_old", false, GHOSTED))
+    _residual_old(_nl.addVector("residual_old", false, GHOSTED)),
```
Could we use a parallel vector instead of a ghosted vector for residual?

YaqiWang replied:
Good point, we possibly can I guess.

loganharbour commented:
@lindsayad I'd appreciate a look from you as well as we spoke the most about this.

fdkong commented on framework/src/timesteppers/DT2.C:
```diff
@@ -229,23 +231,24 @@ DT2::rejectStep()
+  aux.solutionOld().close();
+  aux.solutionOlder().close();
 }
 
```
heart break, lol. So many `close`s

fdkong commented on test/src/auxkernels/GetElementalValueAux.C:
```diff
@@ -34,6 +34,14 @@ GetElementalValueAux::GetElementalValueAux(const InputParameters & params)
+  if (_time_level == "old")
+    _copied_var.slnOld();
+  else if (_time_level == "older")
+    _copied_var.slnOlder();
```
Just out of curiosity, ElementalValueAux is assumed to be related to `time`?

fdkong commented:
> The aux system will always have old and older because we request it in `AuxKernel`. There's room for more optimization here in the future.

In theory, we should be able to get that improved since not  all aux calculations are related to `time`.  It is important for griffin since there are many angle fluxes in aux system if you use NDA. @YaqiWang can confirm this.  However, we could do this optimization in the future. The current PR have already  done a lot of great improvements.

lindsayad commented:
> In theory, we should be able to get that improved since not all aux calculations are related to `time`. It is important for griffin since there are many angle fluxes in aux system if you use NDA. @YaqiWang can confirm this. However, we could do this optimization in the future. The current PR have already done a lot of great improvements.

Agreed that this should wait for another PR.

lindsayad commented on framework/src/systems/SystemBase.C:
```diff
@@ -1351,14 +1359,32 @@ SystemBase::name() const
+std::string
+SystemBase::oldSolutionStateVectorName(const unsigned int state) const
+{
+  mooseAssert(state != 0, "Not an old state");
```
I'm guessing this isn't called very often. If not, I'd make this a `mooseError`

loganharbour replied:
This is a private method. I would prefer the @rwcarlsen philosophy of using assertions for development-based error catching here

lindsayad replied:
K

lindsayad commented on framework/src/systems/SystemBase.C:
```diff
@@ -1367,33 +1393,44 @@ SystemBase::solutionState(const unsigned int state) const
-        _solution_states[i] = &addVector("solution_state_" + std::to_string(i), true, GHOSTED);
-  }
+  else
+    mooseAssert(_solution_states[0] == &solutionInternal(), "Inconsistent current solution");
```
Maybe `mooseAssert` -> `mooseError`

loganharbour replied:
Same thing here - this is checking against an internal API that isn't touched by the user. I'd prefer an assertion if possible

lindsayad commented on framework/src/systems/SystemBase.C:
```diff
@@ -1367,33 +1393,44 @@ SystemBase::solutionState(const unsigned int state) const
+    if (!_solution_states[i])
+      _solution_states[i] = &addVector(oldSolutionStateVectorName(i), true, GHOSTED);
+    else
+      mooseAssert(_solution_states[i] == &getVector(oldSolutionStateVectorName(i)),
```
Maybe `mooseAssert` -> `mooseError`

lindsayad commented on framework/src/variables/MooseVariableData.C:
```diff
@@ -192,6 +192,17 @@ MooseVariableData<OutputType>::MooseVariableData(const MooseVariableFE<OutputTyp
+unsigned int
+MooseVariableData<OutputType>::needSolutionState() const
+{
+  if (_need_u_older || _need_grad_older || _need_second_older || _need_dof_values_older)
+    return 2;
```
Hmm, just looking at the language in this method, maybe I'm more on the @fdkong side of a `MooseEnum`. Unless we truly do have a higher solution state somewhere that doesn't have *any* associated `older[er]*` language

loganharbour replied:
In an ideal world, I would actually have the abstract state design within variable data as well. What do you think about that?

lindsayad replied:
I'm in favor. I would expect it to decrease # LOC

YaqiWang commented on framework/include/systems/SystemBase.h:
```diff
@@ -232,12 +237,19 @@ class SystemBase : public libMesh::ParallelObject, public ConsoleStreamInterface
+  /**
+   * Registers that the solution state \p state is needed.
+   */
+  void needSolutionState(const unsigned int state);
```
Can we use enum for the argument?

loganharbour replied:
Take a look here: https://github.com/idaholab/moose/pull/16363#discussion_r551435071

YaqiWang replied:
duplicated question. I am ok to use unsigned int, just add more comments here like
```
0 - only current solution is needed
1- both current and old solutions are needed
2 - current, old and older solutions are needed
```

loganharbour replied:
Yeah, I definitely agree. I'll make it clearer. thanks!

YaqiWang commented on framework/include/variables/MooseVariableData.h:
```diff
@@ -478,6 +478,11 @@ class MooseVariableData
 
+  /**
+   * The solution states that are needed for this variable data.
+   */
```
This comment is not clear to me. It says `states` but only asking for one single returned number. Does `0` mean current solution only, `1` mean current plus old, etc.?

loganharbour replied:
See https://github.com/idaholab/moose/pull/16363#discussion_r551435479. 

I plan on making this more clear.

YaqiWang replied:
Oh, the one in `MooseVariableFieldBase.h` seems clearer and should be copied here.

YaqiWang commented on framework/src/executioners/EigenExecutionerBase.C:
```diff
@@ -67,6 +67,8 @@ EigenExecutionerBase::EigenExecutionerBase(const InputParameters & parameters)
   // FIXME: currently we have to use old and older solution vectors for power iteration.
   //       We will need 'step' in the future.
   _problem.transient(true);
```
Do we still need this line?

loganharbour replied:
I'm not sure. Will check - but I'm not too concerned about a system that we're phasing out.

YaqiWang commented on framework/src/systems/DisplacedSystem.C:
```diff
@@ -46,22 +42,19 @@ DisplacedSystem::getVector(const std::string & name)
+  if (_sys.have_vector(name))
+    return _sys.get_vector(name);
+  else
+    return _undisplaced_system.getVector(name);
```
Should we error out here instead of returning the vector not for displaced system?

loganharbour replied:
The underlying interface in libmesh will error if not found.

YaqiWang commented on framework/src/systems/NonlinearSystem.C:
```diff
@@ -91,8 +91,8 @@ compute_postcheck(const NumericVector<Number> & old_soln,
-        fe_problem, fe_problem.es().add_system<TransientNonlinearImplicitSystem>(name), name),
-    _transient_sys(fe_problem.es().get_system<TransientNonlinearImplicitSystem>(name)),
+        fe_problem, fe_problem.es().add_system<NonlinearImplicitSystem>(name), name),
+    _nl_implicit_sys(fe_problem.es().get_system<NonlinearImplicitSystem>(name)),
```
So using `System` will be in the near future?

loganharbour replied:
Correct! That one's on @fdkong :)

loganharbour committed:
Rename needSolutionState() to oldestSolutionStateRequested and add docstrings

refs #16339

loganharbour committed:
Resrict AuxKernels to only the current variable value by default

refs #16339

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/654466/) on 4eb138f wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/16363/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/16363/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format bbd2e27f25001c4834201d1d3ff5bb4fae46c84e`

loganharbour commented:
Good for review again.

Last commit is new. Removed the default in AuxKernels for requesting the old/older values and replaced it with a simple interface.

loganharbour commented:
Temp fixes for `_u_old`/`_u_older` in AuxKernels:

- blackbear: https://github.com/idaholab/blackbear/pull/162

friedmud commented on framework/src/auxkernels/VariableTimeIntegrationAux.C:
```diff
@@ -28,7 +28,9 @@ VariableTimeIntegrationAux::validParams()
-    _order(getParam<unsigned int>("order"))
+    _order(getParam<unsigned int>("order")),
+    _u_old(_order != 3 ? uOld() : genericZeroValue<false>()),
+    _u_older(_order == 3 ? uOlder() : genericZeroValue<false>())
```
Hmmm - is there no better way?  It would be great to avoid the ternary.  Maybe it would be better to hide this in the interface underneath `uOld()`?

loganharbour replied:
This is specific to this kernel only. Don't need the older soln if the order is < 3.

I could make it a pointer...

What I was more concerned about was the `uOld()` and `uOlder()` interface.

friedmud replied:
Oh - I see what you're saying.  It looks good then.

friedmud reviewed (changes requested):
Just the one comment for now

fdkong commented:
While this PR might take a while, I have a temporary bandage (at https://github.com/libMesh/libmesh/pull/2832) to avoid allocating memory for petsc matrices when using JFNK .

loganharbour commented:
This is good now. All of the patches are in.

Some of the applications that we update manually will require updates (those that use bison and griffin).

lindsayad reviewed (dismissed):
Do you have a patch for the Rat failure?

loganharbour commented:
@lindsayad can I get a re-review on this? I cherry-picked in the CartesianMeshGenerator fixes so it'll go green.

And for this:

> Do you have a patch for the Rat failure?

We've stopped testing Rat so we're good, now.

fdkong commented:
I will have an extra review today

fdkong reviewed (commented):
We did not move `TransientEigenSystem` to `EigenSystem` ?

    ```
    eigen_problem, eigen_problem.es().add_system<TransientEigenSystem>(name), name),
    _transient_sys(eigen_problem.es().get_system<TransientEigenSystem>(name)),
```

loganharbour commented:
> We did not move `TransientEigenSystem` to `EigenSystem` ?

Darn. Missed that. I'll do that in a separate PR right now.
